plugins/deployment_manager/Spider.ts

```diff
@@ -1,211 +1,265 @@
 import { Contract } from 'ethers';
-import { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { HardhatRuntimeEnvironment as HRE } from 'hardhat/types';
 
 import { Cache } from './Cache';
 import {
-  AliasTemplate,
+  Ctx,
+  AliasRender,
   RelationConfigMap,
+  RelationConfig,
   RelationInnerConfig,
-  aliasTemplateFromAlias,
+  aliasTemplateKey,
   getFieldKey,
   readAlias,
   readField,
 } from './RelationConfig';
 import { Address, Alias, BuildFile } from './Types';
 import { Aliases } from './Aliases';
 import { Proxies } from './Proxies';
+import { ContractMap } from './ContractMap';
 import { Roots } from './Roots';
-import { asArray, cross, debug, getPrimaryContract, mergeABI } from './Utils';
-import { fetchAndCacheContract } from './Import';
+import { asArray, debug, getEthersContract, mergeABI } from './Utils';
+import { fetchAndCacheContract, readAndCacheContract, readContract } from './Import';
 
-function isValidAddress(address: Address): boolean {
-  return address !== '0x0000000000000000000000000000000000000000';
+export interface Spider {
+  roots: Roots;
+  aliases: Aliases;
+  proxies: Proxies;
 }
 
-async function getDiscoverNodes(
+interface Build {
+  buildFile: BuildFile;
+  contract: Contract;
+}
+
+interface DiscoverNode {
+  address: Address;
+  aliasRender: AliasRender;
+  path: Contract[];
+}
+
+function maybeStore(alias: Alias, address: Address, into: Aliases | Proxies): boolean {
+  const maybeExists = into.get(alias);
+  if (maybeExists) {
+    if (maybeExists === address) {
+      return false;
+    } else {
+      throw new Error(`Had ${alias} -> ${maybeExists}, not ${address}`);
+    }
+  } else {
+    into.set(alias, address);
+    return true;
+  }
+}
+
+async function discoverNodes(
+  path: Contract[],
   contract: Contract,
-  relationConfig: RelationInnerConfig,
-  alias: Alias
-): Promise<[Address, AliasTemplate][]> {
-  let addresses: string[] = await readField(contract, getFieldKey(alias, relationConfig));
-  let aliasTemplates: AliasTemplate[] = relationConfig.alias
-    ? asArray<AliasTemplate>(relationConfig.alias)
-    : [aliasTemplateFromAlias(alias)];
-
-  return cross(addresses, aliasTemplates);
+  context: Ctx,
+  config: RelationInnerConfig,
+  defaultKeyAndTemplate: string
+): Promise<DiscoverNode[]> {
+  const addresses = await readField(contract, getFieldKey(config, defaultKeyAndTemplate), context);
+  const templates = config.alias ? asArray(config.alias) : [defaultKeyAndTemplate];
+  return addresses.map((address, i) => ({
+    address,
+    aliasRender: { template: templates[i % templates.length], i },
```

Comment:
I'm a little confused with this one.

Reply:
The fact that its modding? This is just for the case where the array of alias templates is a different length than the field addresses. If there's a single alias template, we already relied on using that for all of them, which this still supports, it just now treats the alias list as a cycle. We could probably just throw an error if they are different lengths and aliases is not length 1, but I figured the cycle is less code and *could* actually be useful theoretically

Reply:
Yeah, but I thought before we did a cross product, versus a cycle-list. I think this changes behavior pretty substantially, though maybe for the better.

Reply:
I think the cross product was just wrong? Commented about this in the discord channel and didn't get a response, but pair-wise wouldn't make sense right?

Reply:
If I recall correctly, and it's been a few months, but the idea was that you could provide multiple _ways_ of aliasing something, e.g. by fixed name (like "baseToken") _and_ by something like `asset.symbol()`. So in that case, cross-product made sense. I guess my question is: did we diverge in how we write `relations.ts` that makes zip more appropriate than cross?

Reply:
I think `Nx1` cross makes sense but that's kind of a degen case and also covered by the cycle behavior if you have e.g. array of `N` addresses from `field` with `1` function for computing `alias`